### DISCUSSION TITLE
k3s cluster can not be started

### DISCUSSION THREAD

**Discussion by rmammadli** [k3s-io/k3s#7034] [Q&A]
**Environmental Info:**
K3s Version: 
k3s version v1.23.16+k3s1 (64b0feeb)
go version go1.19.5

Node(s) CPU architecture, OS, and Version: 

Linux workload-cluster-dev-01-cp02-v20221128-0ed10fd2-2n788 5.4.0-132-generic #148-Ubuntu SMP Mon Oct 17 16:02:06 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:

3 servers, 3 agents

**Describe the bug:**

Pods are in 'Pending' state until the taint is removed.

```

ubuntu@workload-cluster-dev-01-cp02-v20221128-0ed10fd2-2n788:~$ sudo k3s kubectl get all -A
NAMESPACE       NAME                                          READY   STATUS             RESTARTS        AGE
cattle-system   pod/cattle-cluster-agent-6f887b9d86-klv99     0/1     CrashLoopBackOff   5 (2m24s ago)   7m36s
cattle-system   pod/kube-api-auth-svjzq                       1/1     Running            0               7m36s
kube-system     pod/coredns-5cfbb9f57c-x4glj                  0/1     Pending            0               7m36s
kube-system     pod/helm-install-traefik-crd-fqbh6            0/1     Pending            0               7m36s
kube-system     pod/helm-install-traefik-lj24p                0/1     Pending            0               7m36s
kube-system     pod/local-path-provisioner-5f8bbd68f9-lksm7   0/1     Pending            0               7m36s
kube-system     pod/metrics-server-65cd754bcd-cztzb           0/1     Pending            0               7m36s

```

**Steps To Reproduce:**

Deploy a k3s cluster using rancher.

**Expected behavior:**

K3s cluster is deployed successfully.

**Actual behavior:**

Cluster stucks in 'Updating' state / can not be deployed until the taint below on cp / etcd / master node is removed.

```
...
  - effect: NoExecute
    key: node-role.kubernetes.io/etcd
...
```
due to the fact that system services / default components for instance coredns, does not tolerate the taint above. 

**Additional context / logs:**

**Comment by brandond** [comment]
That's not a standard taint that any of the packaged components are going to tolerate. Why did you decide to taint your nodes with that?

**Comment by rmammadli** [comment]
@brandond thank you very much for your prompt reply.
Actually this taint comes by default during provisioning k3s cluster using rancher for node with etcd role.

**Reply by brandond** [reply]
Ahh, right. I believe that this is expected behavior given what you've asked Rancher to do. This isn't really an issue with K3s; you've asked Rancher to taint your server nodes to create a dedicated etcd/control-plane pool. When you do that, it is expected that none of the packaged components will be able to run until you add some "worker" nodes that are allowed to host pods. 

If you don't want the taint, you should ask Rancher to also use the "worker" role on the servers.

**Reply by rmammadli** [reply]
you are right, but we would like to keep those roles as cp/etcd and worker separately in order to have the same workflow related to already deployed / provisioned rke/rke2 clusters.
My question would be: Is there any way ( or plan ) to deploy packaged component namely coredns with certain toleration(s)?
If not: does it make sense to disable packaged component and install it with different name using toleration(s) needed? 
https://docs.k3s.io/advanced#auto-deploying-manifests

**Reply by brandond** [reply]
Some of the K3s packaged components do tolerate the `control-plane` and `master` role taints, but not `etcd` as K3s is not usually deployed with that taint. As per the issue you linked below, Rancher shouldn't be adding it - it looks like that's been fixed on the Rancher side in a recent release.

**Reply by rmammadli** [reply]
@brandond thank you very much for your inputs!
Got it, unfortunately it is still not the case in rancher v2.7.1, tested with different k3s releases. Do you mean that it is fixed in upcoming rancher v2.7.2?
Note: but it works with rancher v2.6.9, tested version v1.24.8+k3s1.
I think we can close this discussion. 
Thank you again for your time and your efforts you & your team put in development & support of this product, really appreciate it!

**Comment by rmammadli** [comment]
somehow related to this issue:
https://github.com/rancher/rancher/issues/40016
